### PR TITLE
Have test files specify flags instead of option names

### PR DESCRIPTION
**Pull request by ddfisher**
This is less hacky than the previous solution and allows for non-boolean options to be specified.

### CODE REVIEW

File: mypy/test/testcheck.py [hunk 1]
```diff
@@ -278,22 +278,17 @@ def parse_module(self, program_text: str, incremental: int = 0) -> List[Tuple[st
...
+            flag_list = flags.group(1).split()
+            _, options = process_options(flag_list, require_targets=False)
         else:
-            options.python_version = testcase_pyversion(testcase.file, testcase.name)
+            options = Options()
 
-        if platform_flag:
-            options.platform = platform_flag.group(1)
+        # Allow custom python version to override testcase_pyversion
+        if not flag_list or '--python-version' not in flag_list:
```
Comment (Michael0x2a) [inline mypy/test/testcheck.py]: We should also probably check to see if `-2` and `--py2` are in `flag_list`, since those two flags will also set `options.python_version`.

File: mypy/test/testcheck.py [hunk 2]
```diff
@@ -278,22 +278,17 @@ def parse_module(self, program_text: str, incremental: int = 0) -> List[Tuple[st
...
         if flags:
-            options_to_enable = flags.group(1).split()
-            for opt in options_to_enable:
-                setattr(options, opt, True)
-
-        # Allow custom pyversion comment to override testcase_pyversion
-        if version_flag:
-            options.python_version = parse_version(version_flag.group(1))
+            flag_list = flags.group(1).split()
+            _, options = process_options(flag_list, require_targets=False)
```
Comment (Michael0x2a) [inline mypy/test/testcheck.py]: Just a thought, it'd be pretty nice if there was somehow a way to optionally use the BuildSources instead of discarding it and get rid of the `# cmd -m ` hack -- I've run into a few bugs that are difficult to write test cases for due to differences in how `testcheck.parse_module` creates the BuildSources vs `main.process_options`. 

That's probably out-of-scope for this pull request though.

Comment (ddfisher) [inline mypy/test/testcheck.py]: Yeah, this would be nice to do at some point (and I think this PR opens the door nicely), but I think it'd be a bit complicated.

File: mypy/main.py [hunk 3]
```diff
@@ -259,14 +256,15 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
...
-        parser.error("May only specify one of: module, package, files, or command.")
+    if require_targets:
+        code_methods = sum(bool(c) for c in [special_opts.modules,
+                                            special_opts.command,
+                                            special_opts.package,
+                                            special_opts.files])
+        if code_methods == 0:
+            parser.error("Missing target module, package, files, or command.")
+        elif code_methods > 1:
+            parser.error("May only specify one of: module, package, files, or command.")
```
Comment (gvanrossum) [inline mypy/main.py]: Would I be terribly off-topic if I asked you to also verify that -c or -p are only used once, *or* that they are somehow combined? It's a little sad that we can easily check that you don't use -c foo -p bar but we are silently dropping the first of -c foo -c bar.

File: mypy/test/testcheck.py (hunk 2 above)
Comment (gvanrossum) [inline mypy/test/testcheck.py]: +1, +1; maybe add a TODO?

Comment (Michael0x2a) [inline mypy/test/testcheck.py]: If we're going to add a TODO, maybe also add a check that raises an error if the user actually tries passing in file names/modules?

Maybe do something like:

```
targets, options = process_options(flag_list, require_targets=False)
if len(targets) != 0:
    # Raise error/TODO message here
```

File: mypy/main.py (hunk 3 above)
Comment (ddfisher) [inline mypy/main.py]: Because this diff is pretty long, I'd rather not add unrelated changes to it.  Checking for that isn't completely trivial.

Comment (gvanrossum) [inline mypy/main.py]: Another TODO?

--Guido (mobile)

File: mypy/test/testcheck.py (hunk 2 above)
Comment (ddfisher) [inline mypy/test/testcheck.py]: 👍

File: mypy/main.py (hunk 3 above)
Comment (ddfisher) [inline mypy/main.py]: Oops, meant to say I'd make another diff immediately after this one.

Comment (ddfisher) [inline mypy/main.py]: Okay it turns out supporting multiple `-c`s is actually really easy. >_>
#2067
